### PR TITLE
abseil: add components + cache cmake

### PR DESCRIPTION
**Pull request by SpaceIm**
Specify library name and version:  **abseil/all**

- [x] I've read the [guidelines](https://github.com/conan-io/conan-center-index/blob/master/docs/how_to_add_packages.md) for contributing.
- [x] I've followed the [PEP8](https://www.python.org/dev/peps/pep-0008/) style guides for Python code in the recipes.
- [x] I've used the [latest](https://github.com/conan-io/conan/releases/latest) Conan client version.
- [x] I've tried at least one configuration locally with the
      [conan-center hook](https://github.com/conan-io/hooks.git) activated.

closes https://github.com/conan-io/conan-center-index/issues/2924, closes https://github.com/conan-io/conan-center-index/issues/1697

Suggestions are welcome :smile:

### CODE REVIEW

**Comment by SpaceIm** [comment]
Another option would be to:
- parse `abseil-targets.cmake` in package method (before deleting it obviously)
- generate the json file introducted in this PR
- install this json file (or whatever format) in package folder

**Comment by SpaceIm** [comment]
I can wait conan 1.29.0 in CCI, or maybe using `cmake_find_package` in `test_package` which is less buggy in 1.28.1 I believe.

**Comment by Croydon** [comment]
Did you generate the JSON files? (I really hope so 🙈)
If yes, how?

File: recipes/abseil/all/test_package/CMakeLists.txt [hunk 1]
```diff
@@ -1,13 +1,11 @@
-cmake_minimum_required(VERSION 3.2.0)
+cmake_minimum_required(VERSION 3.1.0)
 project(test_package CXX)
 
-# We set it only for the convenience of calling the executable
-# in the package test function
 include(${CMAKE_BINARY_DIR}/conanbuildinfo.cmake)
-conan_basic_setup()
+conan_basic_setup(TARGETS)
```
Comment (prince-chrismc) [inline recipes/abseil/all/test_package/CMakeLists.txt:5]:
```suggestion
conan_basic_setup()
```

File: recipes/abseil/all/conanfile.py [hunk 2]
```diff
@@ -15,71 +16,48 @@ class ConanRecipe(ConanFile):
...
+        self._cmake.definitions["BUILD_TESTING"] = False
+        self._cmake.configure(source_folder=self._source_subfolder)
+        return self._cmake
 
     def build(self):
+        tools.replace_in_file(
+            os.path.join(self._source_subfolder, "CMakeLists.txt"),
+            "project(absl CXX)", """project(absl CXX)
+include(${CMAKE_BINARY_DIR}/conanbuildinfo.cmake)
+conan_basic_setup()""")
```
Comment (prince-chrismc) [inline recipes/abseil/all/conanfile.py:60]: Could this be avoided with a top level CMakeLists.txt wrapper?

**Comment by prince-chrismc** [comment]
> Another option would be to:
> 
> * parse `abseil-targets.cmake` in package method (before deleting it obviously)
> * generate the json file introducted in this PR
> * install this json file (or whatever format) in package folder

I think keeping the parser with the recipe would make maintenance easier!

**Review by danimtb — APPROVED** [review]

**Comment by Croydon** [comment]
Regarding the new auto-merge feature:

If I would "request changes" now, because I think this should not get merged until we have talked further about the generating of the JSON files - would the bot care about this? Or does the bot exclusively look at the approvals?

**Review by Croydon — CHANGES_REQUESTED** [review]
How to generate the JSON files?

**Comment by danimtb** [comment]
@Croydon we are still kind of testing it (now in production), as the automation is configurable. The usage is very basic right now and it counts the reviews from the team and the community and only "blocks" the merge if the latest commit has a red review from any of the team members. We are still doing some changes until we enable it with the proper configuration. The details will be documented as requested in #2857

**Comment by SpaceIm** [comment]
@Croydon @prince-chrismc
It was partially generated with a small script taking `absl-target.cmake` file as input, but I had to manually fix some parts in generated file (like system libs) and I don't know if input file is consistent across CMake version.

I know that current recipe is far from ideal, but at least it brings all abseil imported targets in cmake_find_package and cmake_find_package_multi. When i have time, I'll try to really automatize creation of this json file at "package" time.

**Comment by prince-chrismc** [comment]
I think I would prefer if you checked it in, god forbid someone else needs it at least we have a starting point. 
If you do not commit the script, adding a new version would be much more complicated for anyone else.

As long as ur script says "FYI i do not work and you'll need to manually fix some paths" I think we will be miles of head of where we are today.

does that sound reasonable?

**Review by SSE4 — APPROVED** [review]

**Comment by mathbunnyru** [comment]
lts_2020_09_23 was released (https://github.com/abseil/abseil-cpp/releases/tag/20200923).

May I ask you to generate the JSON file for the new version as well?
Totally fine, if that will be done in another PR.

**Review by uilianries — APPROVED** [review]

**Comment by uilianries** [comment]
> lts_2020_09_23 was released ([abseil/abseil-cpp@`20200923` (release)](https://github.com/abseil/abseil-cpp/releases/tag/20200923)).
> 
> May I ask you to generate the JSON file for the new version as well?
> Totally fine, if it will be done in another PR.

Another PR please, this one is related to components only.

**Comment by Croydon** [comment]
@SpaceIm Please provide your script somewhere, even if it is just in a GitHub Gist or whatever. Otherwise we pretty much have a bus factor of 1 for adding new abseil versions and we will always have to annoy you 😄

**Comment by SpaceIm** [comment]
@Croydon I'm close to provide a PR with last abseil release, where cmake config file is (more or less naively) parsed in `package` method, installed, and read in `package_info`.

**Comment by mathbunnyru** [comment]
Thanks @SpaceIm, it sounds awesome.

**Comment by SpaceIm** [comment]
https://github.com/conan-io/conan-center-index/pull/3050

**Pull request merged**